flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/LocalSlicingWindowAggOperator.java

```diff
@@ -0,0 +1,188 @@
+    private final long windowInterval;
+    private final WindowBuffer.Factory windowBufferFactory;
+    private final WindowCombineFunction.LocalFactory combinerFactory;
+
+    // ------------------------------------------------------------------------
+
+    /** This is used for emitting elements with a given timestamp. */
+    protected transient TimestampedCollector<RowData> collector;
+
+    /** Flag to prevent duplicate function.close() calls in close() and dispose(). */
+    private transient boolean functionsClosed = false;
+
+    /** current watermark of this operator. */
+    private transient long currentWatermark;
+
+    /** The next watermark to trigger windows. */
+    private transient long nextTriggerWatermark;
+
+    /** A buffer to buffers window data in memory and may flush to output. */
+    private transient WindowBuffer windowBuffer;
+
+    public LocalSlicingWindowAggOperator(
+            RowDataKeySelector keySelector,
+            SliceAssigner sliceAssigner,
+            PagedTypeSerializer<RowData> keySer,
+            AbstractRowDataSerializer<RowData> inputSer,
+            GeneratedNamespaceAggsHandleFunction<Long> genAggsHandler) {
+        this(
+                keySelector,
+                sliceAssigner,
+                new RecordsWindowBuffer.Factory(keySer, inputSer),
+                new LocalAggRecordsCombiner.Factory(genAggsHandler, keySer));
+    }
+
+    public LocalSlicingWindowAggOperator(
+            RowDataKeySelector keySelector,
+            SliceAssigner sliceAssigner,
+            WindowBuffer.Factory windowBufferFactory,
+            WindowCombineFunction.LocalFactory combinerFactory) {
+        this.keySelector = keySelector;
+        this.sliceAssigner = sliceAssigner;
+        this.windowInterval = sliceAssigner.getSliceEndInterval();
+        this.windowBufferFactory = windowBufferFactory;
+        this.combinerFactory = combinerFactory;
+    }
+
+    @Override
+    public void open() throws Exception {
+        super.open();
+        functionsClosed = false;
+
+        collector = new TimestampedCollector<>(output);
+        collector.eraseTimestamp();
+
+        final WindowCombineFunction localCombiner =
+                combinerFactory.create(getRuntimeContext(), collector);
+        this.windowBuffer =
+                windowBufferFactory.create(
+                        getContainingTask(),
+                        getContainingTask().getEnvironment().getMemoryManager(),
+                        computeMemorySize(),
+                        localCombiner);
+    }
+
+    @Override
+    public void processElement(StreamRecord<RowData> element) throws Exception {
+        RowData inputRow = element.getValue();
+        RowData key = keySelector.getKey(inputRow);
+        long sliceEnd = sliceAssigner.assignSliceEnd(inputRow, CLOCK_SERVICE);
+        // may flush to output if buffer is full
+        windowBuffer.addElement(key, sliceEnd, inputRow);
```

Comment:
Currently, we buffer input record directly. We could also buffer acc which may cost less memory. What do you think?

Reply:
Use memory buffer can help to reduce GC problems and avoid OOM.

Reply:
yes, I agree with you at this point. However it does not conflict with buffer what. We could also buffer acc, right?